api/layer/encryption/encryption.go

```diff
@@ -0,0 +1,344 @@
+package encryption
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/hex"
+	errorsStd "errors"
+	"fmt"
+	"io"
+
+	"github.com/minio/sio"
+)
+
+// Params contains encryption key info.
+type Params struct {
+	enabled     bool
+	customerKey []byte
+}
+
+// ObjectEncryption stores parsed object encryption headers.
+type ObjectEncryption struct {
+	Enabled   bool
+	Algorithm string
+	HMACKey   string
+	HMACSalt  string
+}
+
+type encryptedPart struct {
+	size          uint64
+	encryptedSize uint64
+}
+
+// Range stores payload interval.
+type Range struct {
+	Start uint64
+	End   uint64
+}
+
+// Decrypter allows decrypt payload of encrypted object.
+type Decrypter struct {
+	reader      io.Reader
+	decReader   io.Reader
+	parts       []encryptedPart
+	currentPart int
+	encryption  Params
+
+	rangeParam *Range
+
+	partDataRemain  uint64
+	encPartRangeLen uint64
+
+	seqNumber uint64
+	decLen    uint64
+	skipLen   uint64
+
+	ln  uint64
+	off uint64
+}
+
+const (
+	blockSize     = 1 << 16 // 64KB
+	fullBlockSize = blockSize + 32
+	aes256KeySize = 32
+)
+
+// NewParams creates new params to encrypt with provided key.
+func NewParams(key []byte) (Params, error) {
+	var p Params
+	if len(key) != aes256KeySize {
+		return p, fmt.Errorf("invalid key size: %d", len(key))
+	}
+	p.enabled = true
+	p.customerKey = make([]byte, aes256KeySize)
+	copy(p.customerKey, key)
+	return p, nil
+}
+
+// Key returns encryption key as slice.
+func (p Params) Key() []byte {
+	return p.customerKey[:]
+}
+
+// Enabled returns true if key isn't empty.
```

Comment:
If we `enabled` depends on whether key is not empty, maybe we can drop `enable` and check in `Enable()` length of key?